File: TestPersonne.py
class Personne:
    def __init__(self,nom,prenom):
        self.nom=nom
        self.prenom=prenom
    def __str__(self):
        return self.nom+" "+self.prenom

class Etudiant(Personne):
    NB=0;
    def __init__(self,nom,prenom,num):
        super().__init__(nom,prenom)
        self.num=num
        Etudiant.NB+=1
    def __str__(self):
        return super().__str__()+" Student "+str(self.num)
    


if __name__ == '__main__':
    #  TEST 1 : Une instance
    p = Personne("Duchemin","paul");
    print(p)
	
    # TEST 2 : Une collection
    al1 = [Personne("Duchemin"+str(i),"paul"+str(i)) for i in range(1,11)]
    for p in al1: print(p)
	
    # TEST 3 : Polymorphisme
    al2 =[]
    for i in range(1,10):
        if i%2==0:
            al2.append(Personne("Duchemin"+str(i),"paul"+str(i)))
        else:
            al2.append(Etudiant("Durand"+str(i),"jules"+str(i) , i))
    for p in al2: print(p)
    
    print("Nb etudiants "+str(Etudiant.NB))

File: TestPersonne.rb
#!/usr/bin/env ruby
class Personne
  attr_reader :nom, :prenom # dynamic getter

  def initialize(nom:, prenom:)
    @nom = nom
    @prenom = prenom
  end

  def to_s
    "#{nom} #{prenom}"
  end
end

class Etudiant < Personne
  attr_reader :num # dynamic getter
  class << self; attr_accessor :nb_etudiant end
  @nb_etudiant = 0

  def initialize(nom:, prenom:, num:)
    super(nom: nom, prenom: prenom) # call implicit parent arguments
    @num = num
    self.class.nb_etudiant += 1 # ou Etudiant.nb_etudiant += 1
  end

  def to_s
    super + " Student #{num}"
  end
end

# Une instance
p = Personne.new(nom: 'Duchemin ', prenom: 'paul')
puts p

# Une Collection
al1 = (1...10).map { |i| Personne.new(nom: "Duchemin#{i}", prenom: "paul#{i}") }
puts al1

# Polymorphisme
al2 = (1...10).map do |i|
  i.even? ? Personne.new(nom: "Duchemin#{i}", prenom: "Paul#{i}")
          : Etudiant.new(nom: "Durand#{i}", prenom: "jules#{i}", num: i)
end
puts al2; puts "Nombre d'étudiants: #{Etudiant.nb_etudiant}"
